pkg/apis/resource/types.go

```diff
@@ -128,13 +128,39 @@ type ResourceSliceSpec struct {
 	// +oneOf=NodeSelection
 	AllNodes bool
 
-	// Devices lists some or all of the devices in this pool.
+	// Devices lists some or all of the devices in this pool. Devices
+	// listed here do not inherit shared attributes from the slice.
```

Comment:
These sorts of caveats are what make Service so hard to reason about.  Why NOT allow it to inherit?  In fact, why not add shared attributes to the "real" PR?

Reply:
Yeah, I agree, if we put attributes at the slice level, *everything* in the slice should inherit them.

One reason to hold off on doing that now is that we haven't figured out if/how we want to do device shapes, as described in options 4, 6, 7 [here](https://github.com/kubernetes-sigs/wg-device-management/issues/20).

Reply:
> Why NOT allow it to inherit? In fact, why not add shared attributes to the "real" PR?

You have to think of this as the situation that we will be in after having released 1.31 without inheritance. We could include shared attributes now, but suppose we don't: how do we evolve this API in the future?

We know that we need to extend it. Both the "inherit attributes" and "partitionable devices" feature are coming, and we can't just add both in 1.31 (timing, API not clear yet) and then never extend this API again.

This PR is about how to do such extensions, not about the specific extension.

Reply:
> Why NOT allow it to inherit?

I didn't really answer that. Suppose we released 1.31 without this feature. In 1.32 we add it and allow all devices to inherit it. Client based on in 1.31 then won't implement the inheritance, so producers don't get what they asked for. This "don't inherit" comment is a warning for producers that they cannot rely on inheritance when using the `Device` field.

Reply:
These are all good objections. I agree that extending the field is better, but then we need a version field or a feature slice (see https://github.com/kubernetes/kubernetes/pull/125953).

Reply:
If we are 75% confident that slice-level shared attributes are coming, that they make YAML smaller, and that we understand the implications (e.g. wrt max number of attributes), then adding them NOW dodges the evolution problem entirely.  Ditto for claim requests, maybe - could we add them now and enforce that they are empty?

That said, I am not sure this evolution problem is so dire.  Compatibility means that things which used to work keep working.

Assume we ship without slice-shared-inherited-attributes in 31.  People build drivers and scheduling around it and issue claims.  Then we add inherited attributes in 32. 
 Exisiting slices and claims keep working.  Now you update a driver, but not the scheduler.  Driver moves a bunch of attributes from per-device to per-slice.  The old scheduler won't see those attributes and will stop resolving claims.  I feel like there are always going to be such component-skew issues as APIs evolve, and it's not SO bad -- it didn't allocate wrong devices, it just didn't allocate.  

Still, if we think it is highly likely to be needed, we should think about adding it, if we can.

Reply:
I agree that this shared attributes aspect could be handled like this.

But the same is not true for partitionable devices, at least the way I understand them right now. To an old scheduler it'll look like it can simply allocate a device, but in reality it is ignoring some additional constraint that is defined in the updated device instance.. That would be wrong.

> Still, if we think it is highly likely to be needed, we should think about adding it, if we can.

I think John and Kevin intend to do it differently than describe in this PR.